File: blaze_flutter_sdk/android/src/main/kotlin/com/blaze/flutterblazesdk/parsers/DataSourceParser.kt
package com.blaze.flutterblazesdk.parsers

import com.blaze.blazesdk.data_source.BlazeAdvancedOrderType
import com.blaze.blazesdk.data_source.BlazeCompositeDataSourceConfig
import com.blaze.blazesdk.data_source.BlazeCompositeDataSourceEntry
import com.blaze.blazesdk.data_source.BlazeDataSourcePersonalizedType
import com.blaze.blazesdk.data_source.BlazeDataSourceType
import com.blaze.blazesdk.data_source.BlazeOrderType
import com.blaze.blazesdk.data_source.BlazeRecommendationsType
import com.blaze.blazesdk.data_source.BlazeWidgetLabel
import com.blaze.blazesdk.prefetch.models.BlazeCachingLevel
import com.blaze.blazesdk.shared.models.BlazeLayoutDirection

/** Extension functions to parse BlazeDataSourceType from Flutter's Map representation */
fun Map<String, Any?>.toBlazeDataSourceType(): BlazeDataSourceType? {
    return when {
        // Search carries an optional "labels" key too, so it must be matched first.
        containsKey("searchText") -> parseSearchDataSource()
        containsKey("labels") -> parseLabelsDataSource()
        containsKey("ids") -> parseIdsDataSource()
        containsKey("recommendationsType") -> parseRecommendationsDataSource()
        containsKey("dataSources") -> parseCompositeDataSource()
        else -> null
    }
}

private fun Map<String, Any?>.parseSearchDataSource(): BlazeDataSourceType.Search? {
    val searchText = get("searchText") as? String ?: return null
    val maxItems = (get("maxItems") as? Number)?.toInt()
    val blazeWidgetLabel = (get("labels") as? Map<String, Any?>)?.toBlazeWidgetLabel()

    return BlazeDataSourceType.Search(
        searchText = searchText,
        maxItems = maxItems,
        blazeWidgetLabel = blazeWidgetLabel
    )
}

private fun Map<String, Any?>.parseLabelsDataSource(): BlazeDataSourceType.Labels? {
    val blazeWidgetLabel =
        (get("labels") as? Map<String, Any?>)?.toBlazeWidgetLabel() ?: return null

    return BlazeDataSourceType.Labels(
        blazeWidgetLabel = blazeWidgetLabel,
        orderType = (get("orderType") as? String)?.toBlazeOrderType(),
        advancedOrderType = (get("advancedOrderType") as? String)?.toBlazeAdvancedOrderType(),
        labelsPriority = parseLabelsPriority(),
        maxItems = (get("maxItems") as? Number)?.toInt(),
        personalizedType =
            (get("personalizedType") as? Map<String, Any?>)?.toBlazePersonalizedType()
    )
}

private fun Map<String, Any?>.parseIdsDataSource(): BlazeDataSourceType.Ids? {
    val ids = get("ids") as? List<*> ?: return null
    val stringIds = ids.filterIsInstance<String>()

    return BlazeDataSourceType.Ids(
        ids = stringIds,
        orderType = (get("orderType") as? String)?.toBlazeOrderType(),
        advancedOrderType = (get("advancedOrderType") as? String)?.toBlazeAdvancedOrderType()
    )
}

private fun Map<String, Any?>.parseRecommendationsDataSource():
        BlazeDataSourceType.Recommendations? {
    val recommendationsMap = get("recommendationsType") as? Map<String, Any?> ?: return null
    val type = recommendationsMap["type"] as? String ?: return null
    val anyLabelFilter =
        (recommendationsMap["anyLabelFilter"] as? List<*>)?.filterIsInstance<String>()
            ?: emptyList()
    val coldStartLabels =
        (recommendationsMap["coldStartLabels"] as? List<*>)?.filterIsInstance<String>()
            ?: emptyList()

    val recommendationsType =
        when (type) {
            "ForYou" -> BlazeRecommendationsType.ForYou(
                anyLabelFilter = anyLabelFilter,
                coldStartLabels = coldStartLabels
            )
            "Trending" -> BlazeRecommendationsType.Trending(anyLabelFilter = anyLabelFilter)
            else -> return null
        }

    return BlazeDataSourceType.Recommendations(type = recommendationsType)
}

/**
 * Parses a [BlazeDataSourceType.Composite] from the "dataSources" key. Nested entries are
 * validated natively (non-empty, no nested composite) - any structural violation surfaces as a
 * [com.blaze.blazesdk.shared.results.BlazeResult.Error] through the normal completion callback,
 * so an empty/malformed list is intentionally passed through rather than pre-validated here.
 */
private fun Map<String, Any?>.parseCompositeDataSource(): BlazeDataSourceType.Composite? {
    val dataSourcesList = get("dataSources") as? List<*> ?: return null
    val entries = dataSourcesList.map { entry ->
        (entry as? Map<String, Any?>)?.toBlazeCompositeDataSourceEntry() ?: return null
    }

    return BlazeDataSourceType.Composite(dataSources = entries)
}

private fun Map<String, Any?>.toBlazeCompositeDataSourceEntry(): BlazeCompositeDataSourceEntry? {
    val dataSourceMap = get("dataSource") as? Map<String, Any?> ?: return null
    val dataSource = dataSourceMap.toBlazeDataSourceType() ?: return null
    val isMandatory = (get("config") as? Map<String, Any?>)?.get("isMandatory") as? Boolean ?: false

    return BlazeCompositeDataSourceEntry(
        dataSource = dataSource,
        config = BlazeCompositeDataSourceConfig(isMandatory = isMandatory)
    )
}

private fun Map<String, Any?>.parseLabelsPriority(): List<BlazeWidgetLabel>? {
    val labelsPriorityList = get("labelsPriority") as? List<*> ?: return null
    val labels = labelsPriorityList.toBlazeWidgetLabelArray()
    return labels.ifEmpty { null }
}

private fun Map<String, Any?>.toBlazeWidgetLabel(): BlazeWidgetLabel? {
    val value = get("value") as? String ?: return null
    return BlazeWidgetLabel.singleLabel(value)
}

private fun String.toBlazeOrderType(): BlazeOrderType? {
    return when (this) {
        "manual" -> BlazeOrderType.MANUAL
        "recentlyUpdatedFirst" -> BlazeOrderType.RECENTLY_UPDATED_FIRST
        "recentlyUpdatedLast" -> BlazeOrderType.RECENTLY_UPDATED_LAST
        "aToZ" -> BlazeOrderType.A_TO_Z
        "zToA" -> BlazeOrderType.Z_TO_A
        "recentlyCreatedFirst" -> BlazeOrderType.RECENTLY_CREATED_FIRST
        "recentlyCreatedLast" -> BlazeOrderType.RECENTLY_CREATED_LAST
        "startTimeDesc" -> BlazeOrderType.START_TIME_DESC
        "startTimeAsc" -> BlazeOrderType.START_TIME_ASC
        "random" -> BlazeOrderType.RANDOM
        else -> null
    }
}

private fun String.toBlazeAdvancedOrderType(): BlazeAdvancedOrderType? {
    return when (this) {
        "liveFirst" -> BlazeAdvancedOrderType.LiveFirst
        else -> null
    }
}

private fun Map<String, Any?>.toBlazePersonalizedType(): BlazeDataSourcePersonalizedType? {
    val type = get("type") as? String ?: return null

    return when (type) {
        "Ids" -> parsePersonalizedTypeIds()
        "Labels" -> parsePersonalizedTypeLabels()
        else -> null
    }
}

private fun Map<String, Any?>.parsePersonalizedTypeIds(): BlazeDataSourcePersonalizedType.Ids? {
    val idsMap = get("ids") as? Map<String, Any?> ?: return null
    val convertedMap = mutableMapOf<BlazeDataSourcePersonalizedType.Ids.ContentType, List<String>>()

    idsMap.forEach { (key, value) ->
        val contentType = key.toPersonalizedContentType() ?: return@forEach
        val ids = (value as? List<*>)?.filterIsInstance<String>() ?: return@forEach
        convertedMap[contentType] = ids
    }

    return BlazeDataSourcePersonalizedType.Ids(idsMap = convertedMap)
}

private fun Map<String, Any?>.parsePersonalizedTypeLabels():
        BlazeDataSourcePersonalizedType.Labels? {
    val labelsFilter =
        (get("labelsFilter") as? Map<String, Any?>)?.toBlazeWidgetLabel() ?: return null

    val labelsPriority =
        (get("labelsPriority") as? List<*>)?.toBlazeWidgetLabelArray() ?: emptyList()

    return BlazeDataSourcePersonalizedType.Labels(
        labelsFilter = labelsFilter,
        labelsPriority = labelsPriority
    )
}

private fun String.toPersonalizedContentType(): BlazeDataSourcePersonalizedType.Ids.ContentType? {
    return when (this) {
        "players" -> BlazeDataSourcePersonalizedType.Ids.ContentType.PLAYERS
        "teams" -> BlazeDataSourcePersonalizedType.Ids.ContentType.TEAMS
        else -> null
    }
}

/** Extension function to convert a List<*> to List<BlazeWidgetLabel> */
private fun List<*>.toBlazeWidgetLabelArray(): List<BlazeWidgetLabel> {
    return this.mapNotNull { item -> (item as? Map<String, Any?>)?.toBlazeWidgetLabel() }
}

/** Extension function to convert Flutter enum name to BlazeCachingLevel */
fun String?.asCachingLevel(): BlazeCachingLevel {
    return when (this) {
        "low" -> BlazeCachingLevel.LOW
        "defaultLevel" -> BlazeCachingLevel.DEFAULT
        "high" -> BlazeCachingLevel.HIGH
        "extreme" -> BlazeCachingLevel.EXTREME
        else -> BlazeCachingLevel.DEFAULT
    }
}

/**
 * Extension function to convert a Flutter enum name to [BlazeLayoutDirection].
 * Returns null when absent or unrecognized, which leaves the native default
 * (follow the system layout direction) in place.
 */
fun String?.asBlazeLayoutDirection(): BlazeLayoutDirection? {
    return when (this) {
        "ltr" -> BlazeLayoutDirection.LTR
        "rtl" -> BlazeLayoutDirection.RTL
        else -> null
    }
}

File: blaze_flutter_sdk/ios/blaze_flutter_sdk/Sources/Parsers/DataSourceParser.swift
import Foundation
import BlazeSDK

extension Dictionary where Key == String, Value == AnyHashable {
    
    /// Converts Flutter's dictionary representation to BlazeDataSourceType
    var toBlazeDataSourceType: BlazeDataSourceType? {
        // Search carries an optional "labels" key too, so it must be matched first.
        if self["searchText"] != nil {
            return parseSearchDataSource()
        } else if self["labels"] != nil {
            return parseLabelsDataSource()
        } else if self["ids"] != nil {
            return parseIdsDataSource()
        } else if self["recommendationsType"] != nil {
            return parseRecommendationsDataSource()
        } else if self["dataSources"] != nil {
            return parseCompositeDataSource()
        }
        return nil
    }

    private func parseSearchDataSource() -> BlazeDataSourceType? {
        guard let searchText = self["searchText"] as? String else {
            return nil
        }

        let maxItems = self["maxItems"] as? Int
        let labels = self["labels"]?.asBlazeWidgetLabel

        return .search(searchText: searchText, maxItems: maxItems, labels: labels)
    }
    
    private func parseLabelsDataSource() -> BlazeDataSourceType? {
        guard let blazeWidgetLabel = self["labels"]?.asBlazeWidgetLabel else {
            return nil
        }
        
        return .labels(
            blazeWidgetLabel,
            labelsPriority: parseLabelsPriority(),
            orderType: (self["orderType"] as? String)?.toBlazeOrderType,
            advancedOrderType: (self["advancedOrderType"] as? String)?.toBlazeAdvancedOrderType,
            maxItems: self["maxItems"] as? Int,
            personalizedType: (self["personalizedType"] as? [String: AnyHashable])?.toBlazePersonalizedType
        )
    }
    
    private func parseIdsDataSource() -> BlazeDataSourceType? {
        guard let ids = self["ids"] as? [String] else {
            return nil
        }
        
        return .ids(
            ids,
            orderType: (self["orderType"] as? String)?.toBlazeOrderType,
            advancedOrderType: (self["advancedOrderType"] as? String)?.toBlazeAdvancedOrderType
        )
    }
    
    private func parseRecommendationsDataSource() -> BlazeDataSourceType? {
        guard let recommendationsDict = self["recommendationsType"] as? [String: AnyHashable],
              let type = recommendationsDict["type"] as? String else {
            return nil
        }
        
        let anyLabelFilter = recommendationsDict["anyLabelFilter"] as? [String] ?? []
        let coldStartLabels = recommendationsDict["coldStartLabels"] as? [String] ?? []

        let recommendationsType: BlazeRecommendationsType
        switch type {
        case "ForYou":
            recommendationsType = .forYou(anyLabelFilter: anyLabelFilter, coldStartLabels: coldStartLabels)
        case "Trending":
            recommendationsType = .trending(anyLabelFilter: anyLabelFilter)
        default:
            return nil
        }

        return .recommendations(recommendationsType)
    }

    /// Nested entries are validated natively (non-empty, no nested composite) - any structural
    /// violation surfaces as a `BlazeResult.failure` through the normal completion callback, so
    /// an empty/malformed list is intentionally passed through rather than pre-validated here.
    private func parseCompositeDataSource() -> BlazeDataSourceType? {
        guard let dataSourcesList = self["dataSources"] as? [AnyHashable] else {
            return nil
        }

        var entries: [BlazeCompositeDataSourceEntry] = []
        for item in dataSourcesList {
            guard let entryDict = item as? [String: AnyHashable],
                  let entry = entryDict.toBlazeCompositeDataSourceEntry else {
                return nil
            }
            entries.append(entry)
        }

        return .composite(dataSources: entries)
    }

    private var toBlazeCompositeDataSourceEntry: BlazeCompositeDataSourceEntry? {
        guard let dataSourceDict = self["dataSource"] as? [String: AnyHashable],
              let dataSource = dataSourceDict.toBlazeDataSourceType else {
            return nil
        }

        let isMandatory = (self["config"] as? [String: AnyHashable])?["isMandatory"] as? Bool ?? false

        return BlazeCompositeDataSourceEntry(
            dataSource: dataSource,
            config: BlazeCompositeDataSourceConfig(isMandatory: isMandatory)
        )
    }

    private func parseLabelsPriority() -> [BlazeWidgetLabel]? {
        guard let labelsPriorityArray = self["labelsPriority"] as? [AnyHashable] else {
            return nil
        }

        let labels = labelsPriorityArray.toBlazeWidgetLabelArray
        return labels.isEmpty ? nil : labels
    }
    

    
    var toBlazePersonalizedType: BlazeDataSourcePersonalizedType? {
        guard let type = self["type"] as? String else {
            return nil
        }
        
        switch type {
        case "Ids":
            return parsePersonalizedTypeIds()
        case "Labels":
            return parsePersonalizedTypeLabels()
        default:
            return nil
        }
    }
    
    private func parsePersonalizedTypeIds() -> BlazeDataSourcePersonalizedType? {
        guard let idsDict = self["ids"] as? [String: [String]] else {
            return nil
        }
        
        var convertedDict: [BlazeDataSourcePersonalizedType.ContentType: [String]] = [:]
        
        for (key, value) in idsDict {
            guard let contentType = key.toPersonalizedContentType else {
                continue
            }
            convertedDict[contentType] = value
        }
        
        return .ids(convertedDict)
    }
    
    private func parsePersonalizedTypeLabels() -> BlazeDataSourcePersonalizedType? {
        guard let labelsFilter = self["labelsFilter"]?.asBlazeWidgetLabel else {
            return nil
        }
        
        let labelsPriorityArray = self["labelsPriority"] as? [AnyHashable] ?? []
        let labelsPriority = labelsPriorityArray.toBlazeWidgetLabelArray
        
        return .labels(
            labelsFilter: labelsFilter,
            labelsPriority: labelsPriority
        )
    }
}

extension AnyHashable {
    
    var asBlazeWidgetLabel: BlazeWidgetLabel? {
        guard let labelDataSource = self as? [String: AnyHashable],
              let labelValue = labelDataSource["value"] as? String else {
            return nil
        }
        return .singleLabel(labelValue)
    }
}

extension Array where Element == AnyHashable {
    
    /// Converts an array of Any objects to an array of BlazeWidgetLabel objects
    var toBlazeWidgetLabelArray: [BlazeWidgetLabel] {
        return self.compactMap { $0.asBlazeWidgetLabel }
    }
}

extension String {
    
    var toBlazeOrderType: BlazeOrderType? {
        switch self {
        case "manual":
            return .manual
        case "recentlyUpdatedFirst":
            return .recentlyUpdatedFirst
        case "recentlyUpdatedLast":
            return .recentlyUpdatedLast
        case "aToZ":
            return .aToZ
        case "zToA":
            return .zToA
        case "recentlyCreatedFirst":
            return .recentlyCreatedFirst
        case "recentlyCreatedLast":
            return .recentlyCreatedLast
        case "startTimeDesc":
            return .startTimeDesc
        case "startTimeAsc":
            return .startTimeAsc
        case "random":
            return .random
        default:
            return nil
        }
    }
    
    var toBlazeAdvancedOrderType: BlazeAdvancedOrderType? {
        switch self {
        case "liveFirst":
            return .liveFirst
        default:
            return nil
        }
    }
    
    var toPersonalizedContentType: BlazeDataSourcePersonalizedType.ContentType? {
        switch self {
        case "players":
            return .players
        case "teams":
            return .teams
        default:
            return nil
        }
    }
    
    var asCachingLevel: BlazeCachePolicyLevel? {
        switch self {
        case "low":
            return .Low
        case "defaultLevel":
            return .Default
        case "high":
            return .High
        case "extreme":
            return .Extreme
        default:
            return nil
        }
    }

    /// Returns nil when absent or unrecognized, which leaves the native default
    /// (follow the system layout direction) in place.
    var asBlazeLayoutDirection: BlazeLayoutDirection? {
        switch self {
        case "ltr":
            return .ltr
        case "rtl":
            return .rtl
        default:
            return nil
        }
    }
    
    var asEntryPointTriggerSource: BlazeEntryPointTriggerSource? {
        switch self {
        case "notification":
            return .notification
        case "deepLink":
            return .deepLink
        case "entryPoint":
            return .entryPoint
        default:
            return nil
        }
    }
    
} 
